include/apq/connection.h

```diff
@@ -1,10 +1,314 @@
+
+#include <apq/detail/bind.h>
+#include <apq/impl/connection.h>
+
 namespace libapq {
 
-class connection {
-public:
-    connection() {}
+using impl::pg_conn_handle;
+
+using no_statistics = decltype(hana::make_map());
+
+/**
+* Marker to tag a connection type.
+*/
+template <typename, typename = std::void_t<>>
+struct is_connection : std::false_type {};
+
+/**
+* This is how connection is determined.
+* Connection is an object for which these next free functions
+* are applicable:
+*       get_connection_oid_map(),
+*       get_connection_socket(),
+*       get_connection_handle()
+*/
+template <typename T>
+struct is_connection<T, std::void_t<
+    decltype(get_connection_oid_map(std::declval<T&>())),
+    decltype(get_connection_socket(std::declval<T&>())),
+    decltype(get_connection_handle(std::declval<T&>()))
+>> : std::true_type {};
+
+
+template <typename, typename = std::void_t<>>
+struct is_connection_wrapper : std::false_type {};
+
+template <typename T>
+struct is_connection_wrapper<::std::shared_ptr<T>> : std::true_type {};
+template <typename T, typename D>
+struct is_connection_wrapper<::std::unique_ptr<T, D>> : std::true_type {};
+template <typename T>
+struct is_connection_wrapper<::std::optional<T>> : std::true_type {};
+template <typename T>
+struct is_connection_wrapper<::boost::shared_ptr<T>> : std::true_type {};
+template <typename T>
+struct is_connection_wrapper<::boost::scoped_ptr<T>> : std::true_type {};
+template <typename T>
+struct is_connection_wrapper<::boost::optional<T>> : std::true_type {};
+
+template <typename T>
+constexpr auto ConnectionWrapper = is_connection_wrapper<std::decay_t<T>>::value;
+
+template <typename T>
+constexpr auto Connection = is_connection<std::decay_t<T>>::value;
+
+template <bool Condition, typename Type = void>
+using Require = std::enable_if_t<Condition, Type>;
+
+
+template <typename T, typename = std::void_t<>>
+struct has_operator_not : std::false_type {};
+template <typename T>
+struct has_operator_not<T, std::void_t<decltype(!std::declval<T>())>>
+    : std::true_type {};
+
+template <typename T>
+constexpr auto OperatorNot = has_operator_not<std::decay_t<T>>::value;
+
+/**
+* Connection type traits
+*/
+template <typename T>
+struct connection_traits {
+    using oid_map = std::decay_t<decltype(get_connection_oid_map(std::declval<T&>()))>;
+    using socket = std::decay_t<decltype(get_connection_socket(std::declval<T&>()))>;
+    using handle = std::decay_t<decltype(get_connection_handle(std::declval<T&>()))>;
+};
+
+/**
+* Function to get connection traits type from object
+*/
+template <typename T, typename = Require<Connection<T>>>
+constexpr connection_traits<std::decay_t<T>> get_connection_traits(T&&) noexcept {return {};}
+
+/**
+* Connection public access functions section 
+*/
+
+/**
+* Overloaded version of unwrap function for Connection.
+* Returns connection object itself.
+*/
+template <typename T>
+inline decltype(auto) unwrap_connection(T&& conn, 
```

Comment:
Why inline modifier is required here?

Reply:
Standard practice to avoid duplicate specialized instantiation in different translation units. Any problems?

Reply:
ok